java/com/google/flatbuffers/FlatBufferBuilder.java

```diff
@@ -247,9 +254,18 @@ public void clear(){
      */
     static ByteBuffer growByteBuffer(ByteBuffer bb, ByteBufferFactory bb_factory) {
         int old_buf_size = bb.capacity();
-        if ((old_buf_size & 0xC0000000) != 0)  // Ensure we don't grow beyond what fits in an int.
+
+        if (old_buf_size == MAX_BUFFER_SIZE) { // Ensure we don't grow beyond what fits in an int.
             throw new AssertionError("FlatBuffers: cannot grow buffer beyond 2 gigabytes.");
-        int new_buf_size = old_buf_size == 0 ? 1 : old_buf_size << 1;
+        }
+
+        // cap buffer size at 2GB if doubling the old buffer size is a value greater than 2GB
+        int new_buf_size = (old_buf_size & 0xC0000000) != 0 ? MAX_BUFFER_SIZE : old_buf_size << 1;
+
+        if (new_buf_size == 0) {
```

Comment:
would be more readable if this checked `old_buf_size` against `0` like the old code. And this can can happen first, with all the other code in the `else` branch. And maybe a different default buffer size while we're at it, `1` is going to cause at least 6 or so reallocs for almost any buffer.. maybe `256` ?

Reply:
Will do.

Regarding the default buffer size, I believe I see two defaults: one byte [here](https://github.com/google/flatbuffers/blob/b69fc8cc9584436c1bc48012e9c426843ec1a055/java/com/google/flatbuffers/FlatBufferBuilder.java#L77) and in `growByteBuffer`, and 1KiB [here](https://github.com/google/flatbuffers/blob/b69fc8cc9584436c1bc48012e9c426843ec1a055/java/com/google/flatbuffers/FlatBufferBuilder.java#L104). Perhaps should we have 1KiB everywhere?